File: avalon-sandbox/avalon-net/csframework/AvalonFramework/Configuration/AbstractConfiguration.cs
// ============================================================================
//                   The Apache Software License, Version 1.1
// ============================================================================
//
// Copyright (C) 2002-2003 The Apache Software Foundation. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without modifica-
// tion, are permitted provided that the following conditions are met:
//
// 1. Redistributions of  source code must  retain the above copyright  notice,
//    this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice,
//    this list of conditions and the following disclaimer in the documentation
//    and/or other materials provided with the distribution.
//
// 3. The end-user documentation included with the redistribution, if any, must
//    include  the following  acknowledgment:  "This product includes  software
//    developed  by the  Apache Software Foundation  (http://www.apache.org/)."
//    Alternately, this  acknowledgment may  appear in the software itself,  if
//    and wherever such third-party acknowledgments normally appear.
//
// 4. The names "Jakarta", "Avalon", "Excalibur" and "Apache Software Foundation"
//    must not be used to endorse or promote products derived from this  software
//    without  prior written permission. For written permission, please contact
//    apache@apache.org.
//
// 5. Products  derived from this software may not  be called "Apache", nor may
//    "Apache" appear  in their name,  without prior written permission  of the
//    Apache Software Foundation.
//
// THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED WARRANTIES,
// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS  FOR A PARTICULAR  PURPOSE ARE  DISCLAIMED.  IN NO  EVENT SHALL  THE
// APACHE SOFTWARE  FOUNDATION  OR ITS CONTRIBUTORS  BE LIABLE FOR  ANY DIRECT,
// INDIRECT, INCIDENTAL, SPECIAL,  EXEMPLARY, OR CONSEQUENTIAL  DAMAGES (INCLU-
// DING, BUT NOT LIMITED TO, PROCUREMENT  OF SUBSTITUTE GOODS OR SERVICES; LOSS
// OF USE, DATA, OR  PROFITS; OR BUSINESS  INTERRUPTION)  HOWEVER CAUSED AND ON
// ANY  THEORY OF LIABILITY,  WHETHER  IN CONTRACT,  STRICT LIABILITY,  OR TORT
// (INCLUDING  NEGLIGENCE OR  OTHERWISE) ARISING IN  ANY WAY OUT OF THE  USE OF
// THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
// This software  consists of voluntary contributions made  by many individuals
// on  behalf of the Apache Software  Foundation. For more  information on the
// Apache Software Foundation, please see <http://www.apache.org/>.
// ============================================================================

namespace Apache.Avalon.Framework
{
	using System;
	using System.Collections;

	/// <summary>
	/// This is an abstract <see cref="IConfiguration"/> implementation
	/// that deals with methods that can be abstracted away
	/// from underlying implementations.
	/// </summary>
	/// <remarks>
	/// <para><b>AbstractConfiguration</b> makes easier to implementers 
	/// to create a new version of <see cref="IConfiguration"/></para>
	/// </remarks>
	public abstract class AbstractConfiguration : IConfiguration
	{
		private bool readOnly;
		private string name;
		private string location;
		private string val;
		private string ns;
		private string prefix;
		private Hashtable attributes = new Hashtable();
		private ConfigurationCollection children = new ConfigurationCollection();

		/// <summary>
		/// Gets a value indicating whether the <see cref="IConfiguration"/> is read-only.
		/// </summary>
		/// <value>
		/// <see langword="true"/> if the <see cref="IConfiguration"/> is read-only;
		/// otherwise, <see langword="false"/>.
		/// </value>
		public bool IsReadOnly
		{
			get
			{
				return readOnly;
			}
		}

		/// <summary>
		/// Gets the name of the <see cref="IConfiguration"/>.
		/// </summary>
		/// <value>
		/// The Name of the <see cref="IConfiguration"/>.
		/// </value>
		public string Name
		{
			get
			{
				return name;
			}
			set
			{
				CheckReadOnly();

				name = value;
			}
		}

		/// <summary>
		/// Gets a string describing location of the <see cref="IConfiguration"/>.
		/// </summary>
		/// <value>
		/// A String describing location of the <see cref="IConfiguration"/>.
		/// </value>
		public string Location
		{
			get
			{
				return location;
			}
			set
			{
				CheckReadOnly();

				location = value;
			}
		}

		/// <summary>
		/// Gets the value of <see cref="IConfiguration"/>.
		/// </summary>
		/// <value>
		/// The Value of the <see cref="IConfiguration"/>.
		/// </value>
		public string Value
		{
			get
			{
				return val;
			}
			set
			{
				CheckReadOnly();

				val = value;
			}
		}

		/// <summary>
		/// Gets the namespace of the <see cref="IConfiguration"/>.
		/// </summary>
		/// <value>
		/// The Namespace of the <see cref="IConfiguration"/>.
		/// </value>
		public string Namespace
		{
			get
			{
				return ns;
			}

			set
			{
				CheckReadOnly();

				ns = value;
			}
		}

		/// <summary>
		/// Gets the prefix of the <see cref="IConfiguration"/>.
		/// </summary>
		/// <value>
		/// The prefix of the <see cref="IConfiguration"/>.
		/// </value>
		public string Prefix
		{
			get
			{
				return prefix;
			}

			set
			{
				CheckReadOnly();

				prefix = value;
			}
		}


		/// <summary>
		/// Gets all child nodes.
		/// </summary>
		/// <value>The <see cref="ConfigurationCollection"/> of child nodes.</value>
		public ConfigurationCollection Children
		{
			get
			{
				if (children == null)
				{
					children = new ConfigurationCollection();
				}

				return children;
			}

			set
			{
				CheckReadOnly();

				children = value;
			}
		}

		/// <summary>
		/// Gets node attributes.
		/// </summary>
		/// <value>
		/// All attributes of the node.
		/// </value>
		public  IDictionary Attributes
		{
			get
			{
				if (attributes == null)
				{
					attributes = new Hashtable();
				}

				return attributes;
			}

			set
			{
				CheckReadOnly();

				attributes = new Hashtable(value);
			}
		}

		/// <summary>
		///	Gets a <see cref="IConfiguration"/> instance encapsulating the specified
		/// child node.
		/// </summary>
		/// <param name="child">The Name of the child node.</param>
		/// <returns>
		///	The <see cref="IConfiguration"/> instance encapsulating the specified
		///	child node.
		/// </returns>
		public IConfiguration GetChild(string child)
		{
			return GetChild(child, false);
		}

		/// <summary>
		///	Gets a <see cref="IConfiguration"/> instance encapsulating the specified
		/// child node.
		/// </summary>
		/// <param name="child">The Name of the child node.</param>
		/// <param name="createNew">
		///	If <see langword="true"/>, a new <see cref="IConfiguration"/>
		/// will be created and returned if the specified child does not exist.
		/// If <see langword="false"/>, <see langword="null"/> will be returned when the specified
		/// child doesn't exist.
		/// </param>
		/// <returns>
		///	The <see cref="IConfiguration"/> instance encapsulating the specified
		///	child node.
		/// </returns>
		public abstract IConfiguration GetChild(string child, bool createNew);

		/// <summary>
		/// Return an <see cref="ConfigurationCollection"/> of <see cref="IConfiguration"/>
		/// elements containing all node children with the specified name.
		/// </summary>
		/// <param name="name">The Name of the children to get.</param>
		/// <returns>
		/// All node children with the specified name
		/// </returns>
		public abstract ConfigurationCollection GetChildren(string name);

		/// <summary>
		/// Gets the value of the node and converts it
		/// into specified <see cref="System.Type"/>.
		/// </summary>
		/// <param name="type">The <see cref="System.Type"/></param>
		/// <returns>The Value converted into the specified type.</returns>
		/// <exception cref="InvalidCastException">
		/// If the convertion fails, an exception will be thrown.
		/// </exception>
		public object GetValue(Type type)
		{
			return GetValue(type, null);
		}

		/// <summary>
		/// Gets the value of the node and converts it
		/// into specified <see cref="System.Type"/>.
		/// </summary>
		/// <param name="type">The <see cref="System.Type"/></param>
		/// <param name="defaultValue">
		/// The Default value returned if the convertion fails.
		/// </param>
		/// <returns>The Value converted into the specified type.</returns>
		public object GetValue(Type type, object defaultValue)
		{

			return Converter.ChangeType(Value, type, defaultValue);
		}

		/// <summary>
		/// Gets the value of specified attribute and
		/// converts it into specified <see cref="System.Type"/>.
		/// </summary>
		/// <param name="name">The Name of the attribute you ask the value of.</param>
		/// <param name="type">The <see cref="System.Type"/></param>
		/// <returns>The Value converted into the specified type.</returns>
		/// <exception cref="InvalidCastException">
		/// If the convertion fails, an exception will be thrown.
		/// </exception>
		public object GetAttribute(string name, Type type)
		{

			return GetAttribute(name, type, null);
		}

		/// <summary>
		/// Gets the value of specified attribute and
		/// converts it into specified <see cref="System.Type"/>.
		/// </summary>
		/// <param name="name">The Name of the attribute you ask the value of.</param>
		/// <param name="type">The <see cref="System.Type"/></param>
		/// <param name="defaultValue">
		/// The Default value returned if the convertion fails.
		/// </param>
		/// <returns>The Value converted into the specified type.</returns>
		public object GetAttribute(string name, Type type, object defaultValue)
		{

			return Converter.ChangeType(Attributes[name], type, defaultValue);
		}

		/// <summary>
		/// Make the configuration read only.
		/// </summary>
		public void MakeReadOnly()
		{
			readOnly = true;
		}

		/// <summary>
		/// Check whether this node is readonly or not.
		/// </summary>
		/// <exception cref="ConfigurationException">
		/// If this node is readonly then an exception will be thrown.
		/// </exception>
		protected void CheckReadOnly()
		{
			if( IsReadOnly )
			{
				throw new ConfigurationException( "Configuration is read only and can not be modified." );
			}
		}
	}
}

File: avalon/framework/src/impl/org/apache/avalon/framework/configuration/AbstractConfiguration.java
/*
 * Copyright (C) The Apache Software Foundation. All rights reserved.
 *
 * This software is published under the terms of the Apache Software License
 * version 1.1, a copy of which has been included with this distribution in
 * the LICENSE file.
 */
package org.apache.avalon.framework.configuration;

import java.util.Iterator;

/**
 * This is an abstract <code>Configuration</code> implementation that deals
 * with methods that can be abstracted away from underlying implementations.
 *
 * @author <a href="mailto:fede@apache.org">Federico Barbieri</a>
 * @author <a href="mailto:stefano@apache.org">Stefano Mazzocchi</a>
 * @author <a href="mailto:donaldp@apache.org">Peter Donald</a>
 * @author <a href="mailto:fumagalli@exoffice.com">Pierpaolo Fumagalli</a>
 * @version CVS $Revision: 1.1 $ $Date: 2001/04/26 14:16:25 $
 */
public abstract class AbstractConfiguration
    implements Configuration
{
    /**
     * Returns the value of the configuration element as an <code>int</code>.
     */
    public int getValueAsInteger()
        throws ConfigurationException
    {
        final String value = getValue();
        try
        {
            if( value.startsWith( "0x" ) )
            {
                return Integer.parseInt( value.substring( 2 ), 16 );
            }
            else if( value.startsWith( "0o" ) )
            {
                return Integer.parseInt( value.substring( 2 ), 8 );
            }
            else if( value.startsWith( "0b" ) )
            {
                return Integer.parseInt( value.substring( 2 ), 2 );
            }
            else
            {
                return Integer.parseInt( value );
            }
        }
        catch( final Exception nfe )
        {
            throw
                new ConfigurationException( "Cannot parse the value of the configuration " +
                                            "element \"" + getName() + "\" as an integer" );
        }
    }

    /**
     * Returns the value of the configuration element as an <code>int</code>.
     */
    public int getValueAsInteger( final int defaultValue )
    {
        try
        {
            return getValueAsInteger();
        }
        catch( final ConfigurationException ce )
        {
            return defaultValue;
        }
    }

    /**
     * Returns the value of the configuration element as a <code>long</code>.
     */
    public long getValueAsLong()
        throws ConfigurationException
    {
        final String value = getValue();
        try
        {
            if( value.startsWith( "0x" ) )
            {
                return Long.parseLong( value.substring( 2 ), 16 );
            }
            else if( value.startsWith( "0o" ) )
            {
                return Long.parseLong( value.substring( 2 ), 8 );
            }
            else if( value.startsWith( "0b" ) )
            {
                return Long.parseLong( value.substring( 2 ), 2 );
            }
            else return Integer.parseInt(value);
        }
        catch( final Exception nfe )
        {
            throw new ConfigurationException( "Cannot parse the value of the " +
                                              "configuration element \"" + getName() +
                                              "\" as a long" );
        }
    }

    /**
     * Returns the value of the configuration element as a <code>long</code>.
     */
    public long getValueAsLong( final long defaultValue )
    {
        try
        {
            return getValueAsLong();
        }
        catch( final ConfigurationException ce )
        {
            return defaultValue;
        }
    }

    /**
     * Returns the value of the configuration element as a <code>float</code>.
     */
    public float getValueAsFloat()
        throws ConfigurationException
    {
        final String value = getValue();
        try
        {
            return Float.parseFloat( value );
        }
        catch( final Exception nfe )
        {
            throw new ConfigurationException( "Cannot parse the value of the " +
                                              "configuration element \"" + getName() +
                                              "\" as a float" );
        }
    }

    /**
     * Returns the value of the configuration element as a <code>float</code>.
     */
    public float getValueAsFloat( final float defaultValue )
    {
        try
        {
            return getValueAsFloat();
        }
        catch( final ConfigurationException ce )
        {
            return(defaultValue);
        }
    }

    /**
     * Returns the value of the configuration element as a <code>boolean</code>.
     */
    public boolean getValueAsBoolean()
        throws ConfigurationException
    {
        final String value = getValue();
        if( value.equals( "true" ) ) return true;
        else if( value.equals( "false" ) ) return false;
        else
        {
            throw new ConfigurationException( "Cannot parse the value of the " +
                                              "configuration element \"" +
                                              getName() + "\" as a boolean" );
        }
    }

    /**
     * Returns the value of the configuration element as a <code>boolean</code>.
     */
    public boolean getValueAsBoolean( final boolean defaultValue )
    {
        try
        {
            return getValueAsBoolean();
        }
        catch( final ConfigurationException ce )
        {
            return defaultValue;
        }
    }

    /**
     * Returns the value of the configuration element as a <code>String</code>.
     */
    public String getValue( final String defaultValue )
    {
        try
        {
            return getValue();
        }
        catch( final ConfigurationException ce )
        {
            return defaultValue;
        }
    }

    /**
     * Returns the value of the attribute specified by its name as an
     * <code>int</code>.
     */
    public int getAttributeAsInteger( final String name )
        throws ConfigurationException
    {
        final String value = getAttribute( name );
        try
        {
            if( value.startsWith( "0x" ) )
            {
                return Integer.parseInt( value.substring( 2 ), 16 );
            }
            else if( value.startsWith( "0o" ) )
            {
                return Integer.parseInt( value.substring( 2 ), 8);
            }
            else if( value.startsWith( "0b" ) )
            {
                return Integer.parseInt( value.substring( 2 ), 2 );
            }
            else
            {
                return Integer.parseInt(value);
            }
        }
        catch( final Exception nfe )
        {
            throw new ConfigurationException( "Cannot parse the value of the attribute \"" +
                                              name + "\" of the configuration element \"" +
                                              getName() + "\" as an integer" );
        }
    }

    /**
     * Returns the value of the attribute specified by its name as an
     * <code>int</code>.
     */
    public int getAttributeAsInteger( final String name, final int defaultValue )
    {
        try
        {
            return getAttributeAsInteger( name );
        }
        catch( final ConfigurationException ce )
        {
            return defaultValue;
        }
    }

    /**
     * Returns the value of the attribute specified by its name as a
     * <code>long</code>.
     */
    public long getAttributeAsLong( final String name )
        throws ConfigurationException
    {
        final String value = getAttribute( name );

        try
        {
            if( value.startsWith( "0x" ) )
            {
                return Long.parseLong( value.substring( 2 ), 16 );
            }
            else if( value.startsWith( "0o" ) )
            {
                return Long.parseLong( value.substring( 2 ), 8 );
            }
            else if( value.startsWith( "0b" ) )
            {
                return Long.parseLong( value.substring( 2 ), 2);
            }
            else
            {
                return Integer.parseInt( value );
            }
        }
        catch( final Exception nfe )
        {
            throw new ConfigurationException( "Cannot parse the value of the attribute \"" +
                                              name + "\" of the configuration element \"" +
                                              getName() + "\" as a long" );
        }
    }

    /**
     * Returns the value of the attribute specified by its name as a
     * <code>long</code>.
     */
    public long getAttributeAsLong( final String name, final long defaultValue )
    {
        try
        {
            return getAttributeAsLong( name );
        }
        catch( final ConfigurationException ce )
        {
            return defaultValue;
        }
    }

    /**
     * Returns the value of the attribute specified by its name as a
     * <code>float</code>.
     */
    public float getAttributeAsFloat( final String name )
        throws ConfigurationException
    {
        final String value = getAttribute( name );
        try
        {
            return Float.parseFloat( value );
        }
        catch( final Exception e )
        {
            throw new ConfigurationException( "Cannot parse the value of the attribute \"" +
                                              name + "\" of the configuration element \"" +
                                              getName() + "\" as a float" );
        }
    }

    /**
     * Returns the value of the attribute specified by its name as a
     * <code>float</code>.
     */
    public float getAttributeAsFloat( final String name, final float defaultValue )
    {
        try
        {
            return getAttributeAsFloat( name );
        }
        catch( final ConfigurationException ce )
        {
            return defaultValue;
        }
    }

    /**
     * Returns the value of the attribute specified by its name as a
     * <code>boolean</code>.
     */
    public boolean getAttributeAsBoolean( final String name )
        throws ConfigurationException
    {
        final String value = getAttribute( name );

        if( value.equals( "true" ) ) return true;
        else if( value.equals( "false" ) ) return false;
        else
        {
            throw new ConfigurationException( "Cannot parse the value of the attribute \"" +
                                              name + "\" of the configuration element \"" +
                                              getName() + "\" as a boolean" );
        }
    }

    /**
     * Returns the value of the attribute specified by its name as a
     * <code>boolean</code>.
     */
    public boolean getAttributeAsBoolean( final String name, final boolean defaultValue )
    {
        try
        {
            return getAttributeAsBoolean( name );
        }
        catch( final ConfigurationException ce )
        {
            return defaultValue;
        }
    }

    /**
     * Returns the value of the attribute specified by its name as a
     * <code>String</code>.
     */
    public String getAttribute( final String name, final String defaultValue )
    {
        try
        {
            return getAttribute( name );
        }
        catch( final ConfigurationException ce )
        {
            return defaultValue;
        }
    }

    /**
     * Return the first <code>Configuration</code> object child of this
     * associated with the given name.
     */
    public Configuration getChild( final String name )
    {
        return getChild( name, true );
    }

    /**
     * Return the first <code>Configuration</code> object child of this
     * associated with the given name.
     */
    public Configuration getChild( final String name, final boolean createNew )
    {
        final Configuration[] children = getChildren( name );
        if( children.length > 0 )
        {
            return children[ 0 ];
        }
        else
        {
            if( createNew )
            {
                return new DefaultConfiguration( name, "-" );
            }
            else
            {
                return null;
            }
        }
    }
}
